In [24]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import scipy.stats as stats
import statsmodels.api as sm
from sklearn import metrics
from sklearn.linear_model import Lasso,LassoCV,Ridge,RidgeCV,LinearRegression,SGDRegressor,LogisticRegression
from sklearn.tree import DecisionTreeRegressor,ExtraTreeRegressor,DecisionTreeClassifier,ExtraTreeClassifier
from sklearn.ensemble import AdaBoostRegressor,BaggingRegressor,GradientBoostingRegressor,RandomForestRegressor
from sklearn.feature_selection import RFE
from sklearn.model_selection import GridSearchCV,KFold,LeaveOneOut,train_test_split
from sklearn.cluster import AgglomerativeClustering,hierarchical,KMeans
from sklearn.naive_bayes import GaussianNB
from sklearn.preprocessing import MinMaxScaler,LabelEncoder
from sklearn.neighbors import KNeighborsClassifier
# !pip install xgboost
#from XGboost import XGBRegressor,XGBClassifier
from mlxtend.feature_selection import SequentialFeatureSelector as sfs
import warnings
warnings.filterwarnings('ignore')
from sklearn.ensemble import RandomForestClassifier,BaggingClassifier,AdaBoostClassifier,GradientBoostingClassifier
from sklearn.ensemble import VotingClassifier




In [25]:
df = pd.read_csv('train.csv').as_matrix()

In [26]:
df

array([[1, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [1, 0, 0, ..., 0, 0, 0],
       ...,
       [7, 0, 0, ..., 0, 0, 0],
       [6, 0, 0, ..., 0, 0, 0],
       [9, 0, 0, ..., 0, 0, 0]], dtype=int64)

In [27]:
df.shape

(42000, 785)

In [28]:
x_train = df[0:29400,1:]

In [29]:
y_train = df[0:29400,0]

In [30]:
x_train.shape

(29400, 784)

In [31]:
y_train.shape 

(29400,)

In [32]:
x_test = df[29400:,1:]

In [33]:
y_test = df[29400:,0]

In [34]:
x_test.shape

(12600, 784)

In [35]:
y_test.shape

(12600,)

In [36]:
x_test_main = pd.read_csv('test.csv')

### Logistic Regression

In [37]:
from sklearn.linear_model import LogisticRegression

In [38]:
lr = LogisticRegression()

In [39]:
model= lr.fit(x_train,y_train)

In [40]:
model.score(x_train,y_train)

0.9418707482993197

In [41]:
y_pred=model.predict(x_test)

##### prdicting model with test data

In [42]:
y_pred=model.predict(x_test_main)

In [43]:
y_pred = pd.DataFrame(y_pred)

In [44]:
y_pred.to_csv('submission_lr.csv',index=False)

In [45]:
from sklearn.metrics import classification_report

In [46]:
#print(classification_report(y_test,y_pred))

In [47]:
from sklearn.metrics import accuracy_score

In [49]:
#accuracy_score(y_test,y_pred)

### DecisionTree Classifier

In [50]:
#GridSearchCV to find optimal max_depth
from sklearn.model_selection import KFold
from sklearn.model_selection import GridSearchCV
#specify number of folds for k-fold CV
n_folds=5
#parameters to build the model on
parameters={'max_depth':range(1,40)}
#instantiate the model
dtree=DecisionTreeClassifier(criterion='entropy',random_state=100)

#fit tree on training data
tree=GridSearchCV(dtree,parameters,cv=n_folds,scoring='accuracy')
tree.fit(x_train,y_train)

GridSearchCV(cv=5, error_score=nan,
             estimator=DecisionTreeClassifier(ccp_alpha=0.0, class_weight=None,
                                              criterion='entropy',
                                              max_depth=None, max_features=None,
                                              max_leaf_nodes=None,
                                              min_impurity_decrease=0.0,
                                              min_impurity_split=None,
                                              min_samples_leaf=1,
                                              min_samples_split=2,
                                              min_weight_fraction_leaf=0.0,
                                              presort='deprecated',
                                              random_state=100,
                                              splitter='best'),
             iid='deprecated', n_jobs=None,
             param_grid={'max_depth': range(1, 40)}, pre_dispatch='2*n_jobs'

In [51]:
#scores of GridSearchCV
scores=tree.cv_results_
pd.DataFrame(scores).head()

mean_fit_time  std_fit_time  mean_score_time  std_score_time  \
0       1.887031      0.305529         0.076756        0.068766   
1       2.604942      0.339898         0.032378        0.005391   
2       2.945268      0.244238         0.034286        0.003256   
3       4.148695      0.183173         0.029984        0.003519   
4       5.216794      0.180333         0.027860        0.007928   

  param_max_depth            params  split0_test_score  split1_test_score  \
0               1  {'max_depth': 1}           0.202551           0.202891   
1               2  {'max_depth': 2}           0.336395           0.312075   
2               3  {'max_depth': 3}           0.502381           0.485884   
3               4  {'max_depth': 4}           0.611565           0.613605   
4               5  {'max_depth': 5}           0.700000           0.672279   

   split2_test_score  split3_test_score  split4_test_score  mean_test_score  \
0           0.202041           0.204252           0.204422         0.203231   
1           0.332313           0.327551           0.329422         0.327551   
2           0.491327           0.501531           0.494218         0.495068   
3           0.613095           0.608844           0.604592         0.610340   
4           0.683333           0.682993           0.677891         0.683299   

   std_test_score  rank_test_score  
0        0.000944               39  
1        0.008293               38  
2        0.006234               37  
3        0.003318               36  
4        0.009270               35

In [52]:
param=tree.best_params_
param

{'max_depth': 12}

#### DecisionTreeClassifier

In [53]:
from sklearn.tree import DecisionTreeClassifier

In [54]:
df_default= DecisionTreeClassifier(max_depth=12,random_state=100)

In [55]:
df_default.fit(x_train,y_train)

DecisionTreeClassifier(ccp_alpha=0.0, class_weight=None, criterion='gini',
                       max_depth=12, max_features=None, max_leaf_nodes=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, presort='deprecated',
                       random_state=100, splitter='best')

In [56]:
from sklearn.metrics import classification_report,confusion_matrix,accuracy_score

In [57]:


y_pred_default=df_default.predict(x_test)

print(classification_report(y_test,y_pred_default))

              precision    recall  f1-score   support

           0       0.91      0.91      0.91      1263
           1       0.90      0.96      0.93      1409
           2       0.84      0.82      0.83      1226
           3       0.81      0.80      0.81      1317
           4       0.85      0.86      0.85      1184
           5       0.79      0.79      0.79      1135
           6       0.91      0.88      0.89      1227
           7       0.87      0.89      0.88      1332
           8       0.82      0.78      0.80      1235
           9       0.83      0.83      0.83      1272

    accuracy                           0.85     12600
   macro avg       0.85      0.85      0.85     12600
weighted avg       0.85      0.85      0.85     12600



In [58]:
print(accuracy_score(y_test,y_pred_default))

0.8541269841269842


In [59]:
y_pred_dt=df_default.predict(x_test_main)

In [60]:
y_pred_dt = pd.DataFrame(y_pred_dt)

In [61]:
y_pred_dt.to_csv('submission_dt.csv',index=False)

### Random Forest Classifier

In [62]:
rfc=RandomForestClassifier(criterion = 'entropy',max_depth=12)

In [63]:
rfc.fit(x_train,y_train)

RandomForestClassifier(bootstrap=True, ccp_alpha=0.0, class_weight=None,
                       criterion='entropy', max_depth=12, max_features='auto',
                       max_leaf_nodes=None, max_samples=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, n_estimators=100,
                       n_jobs=None, oob_score=False, random_state=None,
                       verbose=0, warm_start=False)

In [64]:
rfc.score(x_train,y_train)

0.99578231292517

In [65]:
rfc.feature_importances_

array([0.00000000e+00, 0.00000000e+00, 0.00000000e+00, 0.00000000e+00,
       0.00000000e+00, 0.00000000e+00, 0.00000000e+00, 0.00000000e+00,
       0.00000000e+00, 0.00000000e+00, 0.00000000e+00, 0.00000000e+00,
       0.00000000e+00, 0.00000000e+00, 0.00000000e+00, 0.00000000e+00,
       0.00000000e+00, 0.00000000e+00, 0.00000000e+00, 0.00000000e+00,
       0.00000000e+00, 0.00000000e+00, 0.00000000e+00, 0.00000000e+00,
       0.00000000e+00, 0.00000000e+00, 0.00000000e+00, 0.00000000e+00,
       0.00000000e+00, 0.00000000e+00, 0.00000000e+00, 0.00000000e+00,
       0.00000000e+00, 0.00000000e+00, 0.00000000e+00, 0.00000000e+00,
       1.47232852e-06, 0.00000000e+00, 0.00000000e+00, 6.71332652e-07,
       1.15170585e-06, 6.44460160e-07, 0.00000000e+00, 4.39999122e-06,
       6.59586234e-07, 9.28019565e-07, 0.00000000e+00, 0.00000000e+00,
       0.00000000e+00, 0.00000000e+00, 0.00000000e+00, 0.00000000e+00,
       0.00000000e+00, 0.00000000e+00, 0.00000000e+00, 0.00000000e+00,
      

In [66]:
predictions=rfc.predict(x_test)

In [67]:
predictions=rfc.predict(x_test)

In [68]:
y_pred_rf=rfc.predict(x_test_main)

In [69]:
y_pred_rf = pd.DataFrame(y_pred_rf)

In [70]:
y_pred_rf.to_csv('submission_rf.csv',index=False)

In [71]:
metrics.accuracy_score(y_test,predictions)

0.9587301587301588

In [72]:
metrics.recall_score(y_test,predictions,average = 'weighted')

0.9587301587301588

In [73]:
metrics.f1_score(y_test,predictions,average = 'weighted')

0.9586801152758175

In [74]:
metrics.confusion_matrix(y_test,predictions)

array([[1243,    0,    0,    1,    1,    1,    6,    0,   11,    0],
       [   0, 1394,    4,    4,    2,    1,    1,    1,    2,    0],
       [   9,    1, 1174,    7,    4,    2,    6,    9,   11,    3],
       [   1,    4,   28, 1233,    2,   14,    3,   12,   13,    7],
       [   1,    3,    1,    0, 1127,    1,    8,    0,    7,   36],
       [   4,    4,    3,   20,    2, 1079,    5,    0,   11,    7],
       [   8,    3,    3,    0,    5,    4, 1201,    0,    3,    0],
       [   0,    5,   12,    3,    4,    0,    0, 1285,    4,   19],
       [   1,   14,    6,    8,    4,   11,    6,    1, 1165,   19],
       [   5,    5,    6,   19,   21,    8,    0,   22,    7, 1179]],
      dtype=int64)

In [75]:
metrics.classification_report(y_test,predictions)

'              precision    recall  f1-score   support\n\n           0       0.98      0.98      0.98      1263\n           1       0.97      0.99      0.98      1409\n           2       0.95      0.96      0.95      1226\n           3       0.95      0.94      0.94      1317\n           4       0.96      0.95      0.96      1184\n           5       0.96      0.95      0.96      1135\n           6       0.97      0.98      0.98      1227\n           7       0.97      0.96      0.97      1332\n           8       0.94      0.94      0.94      1235\n           9       0.93      0.93      0.93      1272\n\n    accuracy                           0.96     12600\n   macro avg       0.96      0.96      0.96     12600\nweighted avg       0.96      0.96      0.96     12600\n'

### KNN 

In [76]:
from sklearn.neighbors import KNeighborsClassifier 

In [77]:
knn=KNeighborsClassifier(n_neighbors=3)

In [78]:
knn.fit(x_train,y_train)
prediction=knn.predict(x_test)
print('with KNN (K=3) accuracy is:',knn.score(x_test,y_test))#accuracy

with KNN (K=3) accuracy is: 0.9670634920634921


In [79]:
y_pred_knn=knn.predict(x_test_main)

In [80]:
y_pred_knn = pd.DataFrame(y_pred_knn)

In [81]:
y_pred_knn.to_csv('submission_knn.csv',index=False)

In [82]:
# knn.fit(x_train,y_train)
# prediction=knn.predict(x_test)
# print('with KNN (K=2) accuracy is:',knn.score(x_test,y_test))#accuracy

In [83]:
# neig=np.arange(1,25)
# train_accuracy=[]
# test_accuracy=[]
# #loop over different values of k
# for i,k in enumerate(neig):
#     #k from 1 to 25 (exclude)
#     knn=KNeighborsClassifier(n_neighbors=k)
#     #fit with knn
#     knn.fit(x_train,y_train)
#     #train accuracy
#     train_accuracy.append(knn.score(x_train,y_train))
#     #test_accuracy
#     test_accuracy.append(knn.score(x_test,y_test))
    
# #plot
# plt.figure(figsize=(13,8))
# plt.plot(neig,test_accuracy,label='Testing Accuracy')
# plt.plot(neig,train_accuracy,label='Training Accuracy')
# plt.legend()
# plt.title('value vs accuracy',fontsize=20)
# plt.xlabel('number of neighbours',fontsize=20)
# plt.ylabel('accuracy',fontsize=20)
# plt.xticks(neig)
# plt.grid()
# plt.show()
# print('best accuracy is {} with K = {}'.format(np.max(test_accuracy),1+test_accuracy.index(np.max(test_accuracy))))


### Naive Bayes

In [84]:
from sklearn.naive_bayes import GaussianNB

In [85]:
classifier=GaussianNB()
classifier.fit(x_train,y_train)

GaussianNB(priors=None, var_smoothing=1e-09)

In [86]:
y_pred=classifier.predict(x_test)
y_pred

array([8, 6, 9, ..., 9, 6, 9], dtype=int64)

In [87]:
confusion_matrix(y_test,y_pred)

array([[1129,    2,    4,    5,    1,    3,   69,    1,   32,   17],
       [   0, 1340,    2,    8,    1,    4,   12,    0,   32,   10],
       [ 115,   43,  261,  190,    1,    8,  325,    6,  264,   13],
       [  60,   80,   10,  666,    1,    3,   80,   17,  271,  129],
       [  24,   17,   12,   14,  166,    7,  169,   11,  201,  563],
       [ 132,   46,    7,   44,    5,   72,  101,    0,  619,  109],
       [   8,   15,   10,    1,    0,    3, 1172,    0,   17,    1],
       [   3,    4,    1,   20,    6,    1,    3,  466,   43,  785],
       [  16,  204,    2,   13,    2,   14,   40,    1,  705,  238],
       [   5,   15,    2,    6,    8,    4,    1,   19,   25, 1187]],
      dtype=int64)

In [88]:
from sklearn import metrics
from sklearn.metrics import recall_score ,precision_score,accuracy_score,f1_score,classification_report

In [89]:
print(f1_score(y_test,y_pred,average='weighted'))#f1 score higher means model is working better
print(accuracy_score(y_test,y_pred))

0.5236612778886693
0.5685714285714286


### ensemble methods

#### KNN

In [90]:
from sklearn.ensemble import BaggingClassifier

In [91]:
clf1 = KNeighborsClassifier(n_neighbors=3)

In [92]:
bg = BaggingClassifier(base_estimator=clf1, n_estimators=10, max_samples=0.8, max_features=0.8)

In [93]:
bg.fit(x_train,y_train)

BaggingClassifier(base_estimator=KNeighborsClassifier(algorithm='auto',
                                                      leaf_size=30,
                                                      metric='minkowski',
                                                      metric_params=None,
                                                      n_jobs=None,
                                                      n_neighbors=3, p=2,
                                                      weights='uniform'),
                  bootstrap=True, bootstrap_features=False, max_features=0.8,
                  max_samples=0.8, n_estimators=10, n_jobs=None,
                  oob_score=False, random_state=None, verbose=0,
                  warm_start=False)

In [94]:
y_pred_bg = bg.predict(x_test)

In [95]:
y_pred_bg

array([3, 2, 7, ..., 7, 6, 9], dtype=int64)

In [96]:
confusion_matrix(y_test,y_pred_bg)

array([[1255,    2,    0,    0,    0,    1,    5,    0,    0,    0],
       [   0, 1407,    1,    0,    1,    0,    0,    0,    0,    0],
       [  15,   11, 1171,    2,    1,    3,    0,   21,    1,    1],
       [   0,    7,    7, 1268,    0,   13,    0,    5,   11,    6],
       [   0,   14,    0,    0, 1141,    0,    4,    1,    0,   24],
       [   4,    2,    0,   11,    2, 1096,   14,    0,    1,    5],
       [   5,    2,    1,    0,    2,    6, 1211,    0,    0,    0],
       [   0,   12,    4,    0,    3,    0,    0, 1295,    0,   18],
       [   5,   22,    2,   16,    4,   21,    5,    2, 1144,   14],
       [   8,    2,    2,    9,   16,    3,    0,   29,    3, 1200]],
      dtype=int64)

In [97]:
print(f1_score(y_test,y_pred_bg,average='weighted'))#f1 score higher means model is working better
print(accuracy_score(y_test,y_pred_bg))

0.9672287971587173
0.9673015873015873


In [98]:
y_pred_knnbg=bg.predict(x_test_main)

In [99]:
y_pred_knnbg = pd.DataFrame(y_pred_knnbg)

In [100]:
y_pred_knnbg.to_csv('submission_knnbg.csv',index=False)

#### logistic regression

In [101]:
clf2 = LogisticRegression()

In [102]:
bg1 = BaggingClassifier(base_estimator=clf2, n_estimators=10, max_samples=0.8, max_features=0.8)

In [103]:
bg1.fit(x_train,y_train)

BaggingClassifier(base_estimator=LogisticRegression(C=1.0, class_weight=None,
                                                    dual=False,
                                                    fit_intercept=True,
                                                    intercept_scaling=1,
                                                    l1_ratio=None, max_iter=100,
                                                    multi_class='auto',
                                                    n_jobs=None, penalty='l2',
                                                    random_state=None,
                                                    solver='lbfgs', tol=0.0001,
                                                    verbose=0,
                                                    warm_start=False),
                  bootstrap=True, bootstrap_features=False, max_features=0.8,
                  max_samples=0.8, n_estimators=10, n_jobs=None,
                  oob_score=False, random_state=None,

In [104]:
y_pred_bg1 = bg1.predict(x_test)

In [105]:
confusion_matrix(y_test,y_pred_bg1)

array([[1209,    0,    6,    2,    3,   14,   14,    2,   11,    2],
       [   1, 1373,    5,    5,    1,    4,    1,    4,   13,    2],
       [   7,   11, 1092,   22,   12,    8,   17,   14,   37,    6],
       [   5,    5,   38, 1183,    0,   32,    7,   14,   20,   13],
       [   6,    8,    7,    2, 1099,    3,   10,    4,   11,   34],
       [  15,    6,   15,   45,   14,  956,   21,    3,   46,   14],
       [  12,    2,   16,    1,    7,   13, 1169,    2,    5,    0],
       [   3,    2,   15,   11,   10,    2,    1, 1231,    2,   55],
       [   7,   30,   10,   35,    6,   33,   11,    5, 1084,   14],
       [   6,    4,    7,   21,   32,    7,    0,   39,   14, 1142]],
      dtype=int64)

In [106]:
print(f1_score(y_test,y_pred_bg1,average='weighted'))#f1 score higher means model is working better
print(accuracy_score(y_test,y_pred_bg1))

0.9155169800402868
0.9157142857142857


In [107]:
y_pred_lrbg1=bg1.predict(x_test_main)

In [108]:
y_pred_lrbg1 = pd.DataFrame(y_pred_lrbg1)

In [109]:
y_pred_lrbg1.to_csv('submission_lrbg1.csv',index=False)

#### Naive Bayes

In [110]:
clf3 = GaussianNB()

In [111]:
bg2 = BaggingClassifier(base_estimator=clf3, n_estimators=10, max_samples=0.8, max_features=0.8)

In [112]:
bg2.fit(x_train,y_train)

BaggingClassifier(base_estimator=GaussianNB(priors=None, var_smoothing=1e-09),
                  bootstrap=True, bootstrap_features=False, max_features=0.8,
                  max_samples=0.8, n_estimators=10, n_jobs=None,
                  oob_score=False, random_state=None, verbose=0,
                  warm_start=False)

In [113]:
y_pred_bg2 = bg2.predict(x_test)

In [114]:
print(f1_score(y_test,y_pred_bg2,average='weighted'))#f1 score higher means model is working better
print(accuracy_score(y_test,y_pred_bg2))

0.5391368383963318
0.5806349206349206


## Boosting

In [115]:
from sklearn.ensemble import AdaBoostClassifier

In [116]:
clf4 = DecisionTreeClassifier(criterion='entropy', max_depth=12)

In [117]:
boosting = AdaBoostClassifier(base_estimator=clf4, n_estimators=10)

In [118]:
boosting.fit(x_train,y_train)

AdaBoostClassifier(algorithm='SAMME.R',
                   base_estimator=DecisionTreeClassifier(ccp_alpha=0.0,
                                                         class_weight=None,
                                                         criterion='entropy',
                                                         max_depth=12,
                                                         max_features=None,
                                                         max_leaf_nodes=None,
                                                         min_impurity_decrease=0.0,
                                                         min_impurity_split=None,
                                                         min_samples_leaf=1,
                                                         min_samples_split=2,
                                                         min_weight_fraction_leaf=0.0,
                                                         presort='deprecated',
                      

In [119]:
y_pred_boosting =boosting.predict(x_test)

In [120]:
print(f1_score(y_test,y_pred_boosting,average='weighted'))#f1 score higher means model is working better
print(accuracy_score(y_test,y_pred_boosting))

0.9281977512067379
0.9281746031746032


In [121]:
y_pred_boosting=boosting.predict(x_test_main)

In [122]:
y_pred_boosting = pd.DataFrame(y_pred_boosting)

In [123]:
y_pred_boosting.to_csv('submission_y_pred_boosting.csv',index=False)

##### logis  ==             0.9418707482993197
##### dt     ==               0.8541269841269842
##### rf      ==               0.99578231292517
##### knn    ==             0.9670634920634921
##### knn ense      ==  0.9673015873015873
##### log ense  ==       0.9157142857142857
##### boosting  ==      0.9281746031746032



### these are the accuracy of the different models in training set.